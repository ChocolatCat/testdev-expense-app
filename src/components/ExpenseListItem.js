import React from 'react';
import { Link } from "react-router-dom";

const ExpenseListItem = ({id, description, amount, createdAt, note}) => {
    return (
        <div>
            <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
            <h5>{id}</h5>
            <p>{amount} - {createdAt}</p>
            <small>{note}</small>
        </div>
    );
};

export default ExpenseListItem;