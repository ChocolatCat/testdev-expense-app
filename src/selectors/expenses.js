//import {createSelector} from '@reduxjs/toolkit';
import {isAfter, isBefore, isSameDay} from 'date-fns';

const selectExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = expense.createdAt;
        const startDateMatch = startDate ?  isAfter(createdAtMoment, startDate) || isSameDay(startDate, createdAtMoment) : true;
        const endDateMatch = endDate ? isBefore(createdAtMoment, endDate) || isSameDay(endDate, createdAtMoment) : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt > b.createdAt ? 1 : -1
        }
        if(sortBy === 'amount'){
            return a.amount > b.amount ? 1 : -1;
        }
    });  
};

const selectExpenseById = (expenses, id) => {
    return expenses.find( (expense) => {
        return expense.id === id;
    });
};

const selectExpensesTotal = (expenses) => {
    return expenses
        .map( (expense) => expense.amount )
        .reduce( (sum, value) => sum+value, 0);
}

export {selectExpenses, selectExpenseById, selectExpensesTotal};