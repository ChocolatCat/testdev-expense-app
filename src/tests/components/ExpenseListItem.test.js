/** @jest-environment jsdom */

import React from 'react';
import { render, screen } from '@testing-library/react';
import {BrowserRouter} from 'react-router-dom';
import {expenses} from '../fixtures/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';

it('Should render ExpenseListItem', () => {
    const {asFragment} = render(
        <BrowserRouter>
            <ExpenseListItem key={expenses[1].id} {...expenses[1]}/>
        </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});