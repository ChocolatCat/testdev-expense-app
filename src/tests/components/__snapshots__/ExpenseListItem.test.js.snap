// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListItem 1`] = `
<DocumentFragment>
  <div>
    <a
      href="/edit/2"
    >
      <h3>
        Rent
      </h3>
    </a>
    <h5>
      2
    </h5>
    <p>
      40000 - 1659787200000
    </p>
    <small>
      For august
    </small>
  </div>
</DocumentFragment>
`;
